### PR TITLE
compressed fileio, plus better py2/3 unicode/bytes compatibility

### PR DESCRIPTION
**Pull request by bdewilde**
## Description

Changes:

- refactored `fileio`subpackage
    - moved shared (read/write) functions into separate `fileio/utils.py` module
    - added `open_sesame()` function for opening files regardless of compression, relative/user-home-based paths, and missing intermediate directories
    - added `backports.lzma` dependency to catch Python 2 up to Python 3's stdlib
    - added options for writing json files (matching stdlib's `json.dump()`) that can help save space
    - almost all read/write functions now use `open_sesame()`, enabling seamless fileio for gzip, bz2, and lzma compressed files
    - *NB:* certain file mode / compression pairs simply don't work (this is Python's fault), so users may run into exceptions; for example, Python 2 users can't read or write compressed files in text ('wt', 'rt') mode, only binary ('wb' and 'rb'), and Python 3 users can't read or write json in binary mode
    - significantly expanded test coverage of all of this
- cleaned up deprecated/bad Py2/3 `compat` imports, and added better compat functionality for Py2/3 strings
    - now `unicode_type` used for text data, `bytes_type` for binary data, and `string_types` for when either will do
    - also added `unicode_to_bytes()` and `bytes_to_unicode()` functions, which are superfluous in Py2 but occasionally useful in Py3
- added `compression` param to `TextCorpus.save()` and `.load()` to optionally write metadata json file in compressed form
- moved `fileio.write_conll()` function to `export.doc_to_conll()`, which converts a spacy doc into a ConLL-U formatted string; writing that to disk would require another call to `fileio.write_file()`, but that separation is logically consistent with everything else

## Motivation and Context

I originally had a couple `fileio` functions that enabled compressed reads/writes, but it was hacky and inconsistent. These changes rectify the situation. I was also previously overloading the builtin `str`/`unicode`/`bytes` functions, which was both confusing and potentially bug-inducing. Using `unicode_type`, `bytes_type`, and `string_types` is much clearer.

## How Has This Been Tested?

I updated and fixed a bunch of tests. After _much_ futzing, everything's golden. 🎉 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

hey @danvalente, now that you're back... don't suppose you'd like to take a look? :)

### CODE REVIEW

File: textacy/fileio/read.py [hunk 1]
```diff
@@ -4,31 +4,45 @@
...
 from spacy.tokens.doc import Doc as SpacyDoc
 
-from textacy.compat import bzip_open
+from textacy.fileio import open_sesame
 
 JSON_DECODER = json.JSONDecoder()
 
 
-def read_json(filename, mode='rt', encoding=None, prefix=''):
+def read_file(filepath, mode='rt', encoding=None):
```
Comment (danvalente) [inline textacy/fileio/read.py]: I love that you've abstracted away all details of what type of file.

File: textacy/fileio/utils.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+from textacy import compat
+
+
+def open_sesame(filepath, mode='rt',
+                encoding=None, errors=None, newline=None):
+    """
+    Open file ``filepath``. Compression (if any) is inferred from the file
+    extension ('.gz', '.bz2', or '.xz') and handled automatically; '~', '.',
+    and/or '..' in paths are automatically expanded; if writing to a directory
+    that doesn't exist, all intermediate directories are created automatically,
```
Comment (danvalente) [inline textacy/fileio/utils.py]: Hmm... maybe this should be an option?  Like, what if someone has a typo in the path.  Then they have to delete all the stuff that was accidentally created.

File: textacy/fileio/write.py [hunk 3]
```diff
@@ -3,31 +3,56 @@
...
+from textacy.compat import (bytes_to_unicode, bytes_type,
+                            unicode_to_bytes, unicode_type)
+from textacy.fileio import open_sesame, make_dirs
 
 
-def _make_dirs(filename):
-    head, _ = os.path.split(filename)
-    if head and not os.path.exists(head):
-        os.makedirs(head)
+# TODO: keep this, or no?
```
Comment (danvalente) [inline textacy/fileio/write.py]: Is it used anywhere?

File: textacy/fileio/utils.py (hunk 2 above)
Comment (bdewilde) [inline textacy/fileio/utils.py]: Yeah, I wasn't sure which (automatic or optional) would be best. I can switch it.

File: textacy/fileio/write.py (hunk 3 above)
Comment (bdewilde) [inline textacy/fileio/write.py]: Not currently, no. I might add it in as an option, though, or just leave it as a non-private function.

File: textacy/fileio/utils.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+            is an iterable of dicts; the first element of the tuple is an iterable
+            of items' contents, the second is an iterable of their metadata dicts
+        tuple(iterable(str), iterable(list)): if ``itemwise`` is False and ``items``
+            is an iterable of lists; the first element of the tuple is an iterable
+            of items' contents, the second is an iterable of their metadata lists
+    """
+    if itemwise is True:
+        return ((item.pop(content_field), item) for item in items)
+    else:
+        return unzip(((item.pop(content_field), item) for item in items))
```
Comment (danvalente) [inline textacy/fileio/utils.py]: I think this wins the award for longest docstring-to-code ratio in `textacy`

File: textacy/fileio/utils.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+
+def get_filenames(dirname, match_substr=None, ignore_substr=None,
+                  extension=None, ignore_invisible=True, recursive=False):
+    """
+    Yield the full paths of files on disk under directory ``dirname``, optionally
+    filtering for or against particular substrings or file extensions. and crawling all subdirectories.
+
+    Args:
+        dirname (str): /path/to/dir on disk where files to read are saved
+        match_substr (str, optional): match only files with given substring
```
Comment (danvalente) [inline textacy/fileio/utils.py]: Maybe support regex?  Perhaps, not now, but something to put on the roadmap for a future update

**Comment by danvalente** [comment]
Besides the one optional / automatic change, LGTM.

File: textacy/fileio/utils.py (hunk 5 above)
Comment (bdewilde) [inline textacy/fileio/utils.py]: Hm, good idea, and easy to do. Will add!

File: textacy/fileio/utils.py (hunk 4 above)
Comment (bdewilde) [inline textacy/fileio/utils.py]: ha! it's more complicated to explain than it is to code.

this reminds me: i meant to switch the default `itemwise=True` to `itemwise=False`.

**Pull request merged**